### PR TITLE
feat(jsx/dom): provide jsx-runtime and jsx-dev-runtime via jsx/dom

### PR DESCRIPTION
**Pull request by usualoma**
This PR is a continuation of #1981. The following two commits have been added so far.
* 3a91643a231d6f8e9f898b1ef71373eeac7ebfa4
* d1c9b1bdcfae637c00ab4528aa88d5b461038868

If `"jsxImportSource": "hono/jsx"` is specified, all the stringify code will be included and it will become large, so this can be excluded.

Even if you bundle code that includes Suspense and ErrorBoundary, it will still be able to fit in about 4Kb.

```
$ npx esbuild --minify --minify-syntax --tree-shaking=true --bundle dom2.tsx | wc
       1      98    4256
```

```ts
/** @jsxRuntime automatic @jsxImportSource ./src/jsx/dom */
import { Suspense, ErrorBoundary } from './src/jsx/dom'
import { render } from './src/jsx/dom'
import { use } from './src/jsx/hooks'

const promise = new Promise<number>((resolve) => setTimeout(() => resolve(1), 1000))
const UseComponent = () => {
  const num = use(promise)
  return <div>{num}</div>
}

const rejectedPromise = new Promise<number>((_, reject) => setTimeout(() => reject(1), 1000))
const ErrorComponent = () => {
  const num = use(rejectedPromise)
  return <div>{num}</div>
}

export const App = () => {
  return (
    <div>
      <div>
        <div>
          <Suspense fallback={<div>Loading...</div>}>
            <div>
              <h3>UseComponent</h3>
              <UseComponent />
            </div>
          </Suspense>

          <ErrorBoundary fallback={<div>error</div>}>
            <Suspense fallback={<div>Loading...</div>}>
              <div>
                <h3>ErrorComponent</h3>
                <ErrorComponent />
              </div>
            </Suspense>
          </ErrorBoundary>
        </div>
      </div>
    </div>
  )
}

render(<App />, document.getElementById('root'))
```

### Author should do the followings, if applicable

- [x] Add tests
- [x] Run tests
- [x] `yarn denoify` to generate files for Deno

### CODE REVIEW

**Comment by yusukebe** [comment]
Hi @usualoma 

This is super cooool!

If we are using Vite, we can specify the JSX runtime for the build  in the config file:

```ts
import { defineConfig } from 'vite'
import devServer from '@hono/vite-dev-server'

export default defineConfig(({ mode }) => {
  if (mode === 'client') {
    return {
      build: {
        minify: true,
        rollupOptions: {
          input: ['./app/client.tsx'],
          output: {
            entryFileNames: 'static/client.js'
          }
        }
      },
      esbuild: {
        jsxImportSource: 'hono/jsx/dom' // <=== specify `hono/jsx/dom` for the client
      }
    }
  } else {
    return {
      ssr: {
        noExternal: true
      },
      build: {
        minify: true,
        ssr: './_worker.ts'
      },
      plugins: [
        devServer({
          entry: './app/server.tsx'
        })
      ]
    }
  }
})
```

Then, it will be ultra small:

Screenshot 2024-01-15 at 23 54 55

The counter sample is now 2.3KB from 3.6KB with Brotli compression!

Demo: https://9fe4615e.hono-jsx-front-poc.pages.dev/

If we merge #1981, add this PR to v4!

**Comment by yusukebe** [comment]
Btw, a Preact app is 5.7KB.

Screenshot 2024-01-16 at 0 18 47


https://5436b56a.hono-jsx-front-poc.pages.dev/

**Comment by usualoma** [comment]
Hi, @yusukebe 

Additional testing has been completed.
I am also making changes to the `npm test` and would appreciate a review.

**Comment by yusukebe** [comment]
Hi @usualoma 

Thanks. First, please merge the upstream to resolve the conflict. I'll check it later.

**Comment by usualoma** [comment]
@yusukebe resolved!

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "dist"
   ],
   "scripts": {
-    "test": "tsc --noEmit && vitest --run",
+    "test": "tsc --noEmit && vitest --run && vitest -c .vitest.config/jsx-runtime-dom.ts --run && vitest -c .vitest.config/jsx-runtime-dom.ts --run",
```
Comment (yusukebe) [inline package.json:13]: Maybe it's wrong. Should be:

```
tsc --noEmit && vitest --run && vitest -c .vitest.config/jsx-runtime-default.ts --run && vitest -c .vitest.config/jsx-runtime-dom.ts --run
```

**Comment by yusukebe** [comment]
@usualoma 

Thanks! I've left one comment. Check it.

**Comment by usualoma** [comment]
@yusukebe 
Oops, fix typo in [5ac6b00](https://github.com/honojs/hono/pull/1986/commits/5ac6b00f6e8fadaeaf493087a43d662c708d28eb)

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@usualoma 

Thanks. Merge now!

**Pull request merged**